openmls/src/treesync/node/leaf_node.rs

```diff
@@ -545,123 +571,200 @@ impl LeafNode {
+    pub(crate) fn validate_lifetime(&self) -> Result<&Self, LeafNodeValidationError> {
+        // self.validate_lifetime_range()?; // TODO(openmls/annotations#36)
+        self.validate_lifetime_actuality()?;
+
+        Ok(self)
+    }
+
+    /// Check that lifetime range is acceptable.
+    #[allow(unused)] // TODO(openmls/annotations#36)
+    pub(crate) fn validate_lifetime_range(&self) -> Result<&Self, LeafNodeValidationError> {
+        match self.payload.leaf_node_source {
+            LeafNodeSource::KeyPackage(lifetime) => {
+                if lifetime.has_acceptable_range() {
+                    Ok(self)
+                } else {
+                    Err(LeafNodeValidationError::Lifetime(
+                        LifetimeError::RangeTooBig,
+                    ))
+                }
+            }
+            LeafNodeSource::Update => Ok(self),
+            LeafNodeSource::Commit(_) => Ok(self),
+        }
+    }
+
+    /// Check that current time is between `Lifetime.not_before` and `Lifetime.not_after`.
+    pub(crate) fn validate_lifetime_actuality(&self) -> Result<&Self, LeafNodeValidationError> {
+        match self.payload.leaf_node_source {
+            LeafNodeSource::KeyPackage(lifetime) => {
+                if lifetime.is_valid() {
+                    Ok(self)
+                } else {
+                    Err(LeafNodeValidationError::Lifetime(LifetimeError::NotCurrent))
+                }
+            }
+            LeafNodeSource::Update => Ok(self),
+            LeafNodeSource::Commit(_) => Ok(self),
+        }
+    }
+
+    /// Check that all required capabilities are supported by this leaf node.
     pub(crate) fn validate_required_capabilities<'a>(
         &self,
         required_capabilities: impl Into<Option<&'a RequiredCapabilitiesExtension>>,
-    ) -> Result<(), TreeSyncError> {
+    ) -> Result<&Self, LeafNodeValidationError> {
+        // If the GroupContext has a required_capabilities extension, ...
         if let Some(required_capabilities) = required_capabilities.into() {
+            // ... then the required extensions, ...
             for required_extension in required_capabilities.extensions() {
                 if !self.supports_extension(required_extension) {
-                    return Err(TreeSyncError::UnsupportedExtension);
+                    return Err(LeafNodeValidationError::UnsupportedExtensions);
                 }
             }
+
+            // ... proposals, ...
             for required_proposal in required_capabilities.proposals() {
                 if !self.supports_proposal(required_proposal) {
-                    return Err(TreeSyncError::UnsupportedProposal);
+                    return Err(LeafNodeValidationError::UnsupportedProposals);
                 }
             }
+
+            // ... and credential types MUST be listed in the LeafNode's capabilities field.
+            // TODO(#1200)
```

Comment:
Issue #1200 is closed. Am I missing something here?

Reply:
This was resolved by the rebase.